horovod/common/basics.py

```diff
@@ -30,39 +32,79 @@ def __init__(self, pkg_path, *args):
         self.Sum = self.MPI_LIB_CTYPES.horovod_reduce_op_sum()
         self.Adasum = self.MPI_LIB_CTYPES.horovod_reduce_op_adasum()
 
-    def init(self, comm=None):
+    def init(self, comm=None, process_sets=None):
         """A function that initializes Horovod.
 
         Args:
-          comm: List specifying ranks for the communicator, relative to the MPI_COMM_WORLD
-            communicator OR the MPI communicator to use. Given communicator will be duplicated.
-            If None, Horovod will use MPI_COMM_WORLD Communicator.
+
+          comm: One of these possibilities:
+
+            1) List specifying ranks for the communicator, relative to the MPI_COMM_WORLD
+               communicator
+            2) None: Use all ranks of MPI_COMM_WORLD
+            3) MPI communicator to use. Given communicator will be duplicated.
+            4) List of MPI communicators. The first entry will be duplicated and used as
+               the global Horovod communicator. Horovod process sets will be built
+               corresponding to the remaining entries.
+
+          process_sets: If no explicit MPI communicators are passed, one of these possibilities:
+
+            1) None -- do not initialize any process sets
+            2) List[List[int]] -- initialize process sets containing these Horovod ranks
+            3) "dynamic": do not initialize any process sets now, but set environment
+               HOROVOD_DYNAMIC_PROCESS_SETS=1 .
         """
         if comm is None:
             comm = []
+        elif not isinstance(comm, list):
+            comm = [comm]
+        if process_sets is None:
+            process_sets = []
+        elif isinstance(process_sets, str) and process_sets.lower() == "dynamic":
+            process_sets = []
+            os.environ["HOROVOD_DYNAMIC_PROCESS_SETS"] = "1"
```

Comment:
There is no deep reason to have the env var, but it makes things easier in the unit test script: https://github.com/maxhgerlach/horovod/blob/995fc21d5af269e8bea83fbd2cc415408b9854a2/test/parallel/test_tensorflow.py#L59

Since only the first test to be executed can effectively pass parameters to the background thread via `hvd.init()`, every single test function would need to be adapted to make sure Horovod is initialized properly for the process set tests. That felt a bit invasive to me. Alternatively, all process set tests would need to be moved into a separate file (making sure to share all framework-specific setup logic) and a similar split would be required for the PyTorch and MXNet tests in due time.